### PR TITLE
feat(nn-qgrad): added sgd

### PR DESCRIPTION
**Pull request by LeoBuron**
Added the new sgd optimizer and meta classes for storing quantization schemes by name. These quantization schemes are initialized for each layer. The quantization schemes are used by the optimizer to determine how to round the values. 

The optimizer recursively iterates over the model and stores all quantization schenes of each layer in a big list. For each quantization scheme the parameter and its gradient are stored together with the information of the quantization scheme in the quantization parameter. 

I added inplace versions of the quantization functions in the fixed point folder, which is used by the optimizer. I am not sure how to make sure that the quantization functions need to be an inplace version.

### CODE REVIEW

File: elasticai/creator/nn/quantized_grads/base_modules/linear.py [hunk 1]
```diff
@@ -1,14 +1,19 @@
...
 
 
 class MathOperations(Quantize, Add, MatMul, Protocol): ...
 
 
-class Linear(torch.nn.Linear):
+class Linear(_Linear, QuantizedParameters):
     """This module implements a linear layer.
     The output of the convolution is fake quantized. The weights and bias are fake quantized during initialization.
     To keep the weights quantized use only optimizers that apply a quantized update
```
Comment (glencoe) [inline elasticai/creator/nn/quantized_grads/base_modules/linear.py:19]: Don't you even mean

```
IMPORTANT: use only in-place operators for your quantize implementations
```

**Review by glencoe — CHANGES_REQUESTED** [review]
See comments. The rest looks good to me.

I'm not sure, but i also think adding documentation that explains why you only expect inplace operations would be helpful. I mean we don't really want people or future us to go through the hassle of finding that out again. Maybe a counter example in a test case suffices already.

For what it's worth, we as owners can also bypass the reviews now.

File: elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py [hunk 2]
```diff
@@ -61,6 +62,60 @@ def test_round_determinstic_to_fixed_point():
...
+    _round_to_fixed_point_hte_(x, conf.frac_bits)
+    actual = x
+    expected = torch.Tensor(
+        [
+            -2.0,
+            -2.0,
+            -1.75,
+            -1.75,
+            -1.5,
+            -1.5,
```
[the remark this answers is not in the document]
Comment (LeoBuron) [reply elasticai/creator/nn/quantized_grads/fixed_point/test_quantize_to_fixed_point.py:78]: I agree. But I don't have the time at the moment

File: elasticai/creator/nn/quantized_grads/base_modules/batchnorm2d.py [hunk 3]
```diff
@@ -27,7 +32,7 @@ def __init__(
         device: Any = None,
         dtype: Any = None,
     ) -> None:
-        super().__init__(
+        _BatchNorm2d.__init__(
```
[the remark this answers is not in the document]
Comment (LeoBuron) [reply elasticai/creator/nn/quantized_grads/base_modules/batchnorm2d.py:35]: @glencoe Can you give me a better undestanding of what I need to do. What I tried is the following:

´´´python

class A:
    def __init__(self, a, *args, **kwargs):
        self.a = a
        print(f"Init A")

class B:
    def __init__(self, b, *args, **kwargs):
        self.b = b
        print(f"Init B")


class C(A, B):
    def __init__(self, a, b):
        super().__init__(a=a, b=b)


if __name__ == "__main__":
    c = C(1, 2)
    print(c.a)
    print(c.b)

´´´
